### PR TITLE
Add Prepare and Process Proposal ABCI++ methods

### LINKED ISSUE #623: Create ABCI++ branch with celestia specific changes
We should create a release branch that incorporates all of the necessary celestia specific changes with all of the changes made so far with upstream's ABCI++ branch

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This PR incorporates the two needed ABCI++ methods from upstream.
- cherry-picks the [PrepareProposal](https://github.com/tendermint/tendermint/pull/6544) and [ProcessProposal](https://github.com/tendermint/tendermint/pull/7091) ABCI++ methods
- Modifies both methods to work on tendermint v0.34.x
- Modifies both methods to use `types.Data` instead of `types.Txs` or `[][]byte`

Closes #623

related to [#157](https://github.com/celestiaorg/celestia-app/pull/157)

### CODE REVIEW

**Comment by evan-forbes** [comment]
Still unsure as to why the test_apps script is failing. We've had trouble with it in the past, so I'm not convinced that it is a deal breaker for merging this PR because the e2e is passing, and this test will likely be fixed when we use a stable version of ABCI++.

**Comment by evan-forbes** [comment]
tbc, this PR is a placeholder that allows us to start building features in the app that depend upon these two added methods. When the release version of the cosmos-sdk that uses tendermint v0.36.x comes out, then we can create a version of celestia-core that is based on v0.36.x and use that instead

**Review by liamsi — COMMENTED** [review]
This is pretty cool! Looking forward to testing this with the SDK/our app.

File: proto/tendermint/abci/types.proto [hunk 1]
```diff
@@ -36,7 +36,8 @@ message Request {
     RequestOfferSnapshot      offer_snapshot       = 13;
     RequestLoadSnapshotChunk  load_snapshot_chunk  = 14;
     RequestApplySnapshotChunk apply_snapshot_chunk = 15;
-    RequestPreprocessTxs      preprocess_txs       = 16;
+    RequestPrepareProposal    prepare_proposal     = 16;
+    RequestProcessProposal     process_proposal     = 17;
```
Comment (adlerjohn) [inline proto/tendermint/abci/types.proto:40]:
```suggestion
    RequestProcessProposal    process_proposal     = 17;
```

File: state/mocks/evidence_pool.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Code generated by mockery 2.9.0. DO NOT EDIT.
+// Code generated by mockery 2.7.5. DO NOT EDIT.
```
Comment (adlerjohn) [inline state/mocks/evidence_pool.go:1]: Why is this using an _older_ version of `mockery`???

File: types/tx.go [hunk 3]
```diff
@@ -77,6 +77,27 @@ func (txs Txs) SplitIntoShares() NamespacedShares {
 	return shares
 }
 
+// ToSliceOfBytes converts a Txs to slice of byte slices.
+//
+// NOTE: This method should become obsolete once Txs is switched to [][]byte.
+// ref: #2603
```
Comment (adlerjohn) [inline types/tx.go:83]: Link to actual issue instead of just a number

File: abci/types/application.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
 // to be driven by a blockchain-based replication engine via the ABCI.
 // All methods take a RequestXxx argument and return a ResponseXxx argument,
 // except CheckTx/DeliverTx, which take `tx []byte`, and `Commit`, which takes nothing.
-// nolint:lll // ignore for interface
+// ignore for interface
```
Comment (liamsi) [inline abci/types/application.go:11]: I think this comment was specific for the `nolint:lll`. Can be removed.

**Review by liamsi — APPROVED** [review]
Left one nit. The changes here are straightforwardly pulling from upstream. 

I think it would be important to clarify in the opening comment from which branch(es)/commit(s)/version(s) this was pulled exactly. 

Similarly:
> Modifies both methods to work on tendermint v0.34.x

It would be good if the PR opening comment documented what modifications were necessary here.

File: abci/types/application.go (hunk 4 above)
Comment (liamsi) [inline abci/types/application.go:11]:
```suggestion
```

File: types/tx.go (hunk 3 above)
Comment (evan-forbes) [reply types/tx.go]: link added https://github.com/celestiaorg/celestia-core/pull/631/commits/48e9ebfe23afd28b8bbedb9a77c911698087e325

fwiw, this was inherited from one of the cherry picked PRs tho, and has likely been there for a long time. Agree that we should definitely add links where possible tho

File: state/mocks/evidence_pool.go (hunk 2 above)
Comment (evan-forbes) [reply state/mocks/evidence_pool.go]: idk, I think I just had an older version of mockery installed, good catch! Weirdly enough, I couldn't actually get mockery to work using the provided scripts and the updated version. Simply calling go generate ./... instead does however 

https://github.com/celestiaorg/celestia-core/pull/631/commits/90e4244bcdc771108acd31627cf6331c202ec083

**Pull request merged**